### PR TITLE
Fixing gestures with answer buttons not working

### LINKED ISSUE #8527: [Bug] Gestures for answer buttons don't work
###### Reproduction Steps

1. Enable gestures
2. Set some gesture action to an answer action (e.g. answer button 1, answer button 3, answer recommended, etc.)
3. Go to any deck and perform the gesture.

###### Expected Result
The answer action should be performed after the gesture.

###### Actual Result
The answer action is not performed after the gesture.

###### Debug info
Tested on both physical device and emulator.

Physical device info:
```AnkiDroid Version = 2.15alpha39

Android Version = 10

Manufacturer = samsung

Model = SM-A307GN

Hardware = exynos7904

Webview User Agent = Mozilla/5.0 (Linux; Android 10; SM-A307GN Build/QP1A.190711.020; wv) AppleWebKit/537.36 (KHTML, like Gecko) Version/4.0 Chrome/89.0.4389.105 Mobile Safari/537.36

ACRA UUID = 330c9b2a-f22f-47aa-a891-66352d287395

Scheduler = std

Crash Reports Enabled = false

DatabaseV2 Enabled = true
```

Emulator info:
```
AnkiDroid Version = 2.15alpha39

Android Version = 9

Manufacturer = Google

Model = AOSP on IA Emulator

Hardware = ranchu

Webview User Agent = Mozilla/5.0 (Linux; Android 9; AOSP on IA Emulator Build/PSR1.180720.117; wv) AppleWebKit/537.36 (KHTML, like Gecko) Version/4.0 Chrome/69.0.3497.100 Mobile Safari/537.36

ACRA UUID = d28ce267-ee89-4faf-a3d4-3fc904411482

Scheduler = std

Crash Reports Enabled = false

DatabaseV2 Enabled = true
```

###### Research
The issue only affects the latest alpha release. I was able to pinpoint the commit that creates the issue (https://github.com/ankidroid/Anki-Android/commit/559b3f1e47c1dbdb6395007b31592d311337e003).

- [x] I have read the [support page](https://ankidroid.org/docs/help.html) and am reporting a bug or enhancement request specific to AnkiDroid
- [x] I have checked the [manual](https://ankidroid.org/docs/manual.html) and the [FAQ](https://github.com/ankidroid/Anki-Android/wiki/FAQ) and could not find a solution to my issue
- [x] I have searched for similar existing issues here and on the user forum
- [x] (Optional) I have confirmed the issue is not resolved in the latest alpha release ([instructions](https://docs.ankidroid.org/manual.html#betaTesting))

### LINKED ISSUE #8572: [Bug] Gestures not working since 2.15alpha40
###### Reproduction Steps

1. have gestures turned on
2. choose "recommended answer (green) for swipe left
3.  try to answer card with good (swipe left)

###### Expected Result

card is marked good, next card appears

###### Actual Result

nothing happens

###### Debug info
sorry, immediately uninstalled the alpha, so no debug info.  2.15alpha40 on Android 9

###### Research
*Enter an [x] character to confirm the points below:*

- [ x] I have read the [support page](https://ankidroid.org/docs/help.html) and am reporting a bug or enhancement request specific to AnkiDroid
- [ x] I have checked the [manual](https://ankidroid.org/docs/manual.html) and the [FAQ](https://github.com/ankidroid/Anki-Android/wiki/FAQ) and could not find a solution to my issue
- [ x] I have searched for similar existing issues here and on the user forum
- [x ] (Optional) I have confirmed the issue is not resolved in the latest alpha release ([instructions](https://docs.ankidroid.org/manual.html#betaTesting))

### PR DESCRIPTION
**Pull request by kuroni**
## Purpose / Description
Fixing gestures with answer buttons not working on the latest alpha release.

## Fixes
Fixes #8527 
Fixes #8572.

## Approach
Completely removing the usage of `performClickWithAction`, opting to `answerCard`, since only touching actions on buttons are registered now.

## How Has This Been Tested?

Tested on my physical device.

## Checklist

- [x] You have not changed whitespace unnecessarily (it makes diffs hard to read)
- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] Your code follows the style of the project (e.g. never omit braces in `if` statements) 
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code

### CODE REVIEW

**Review by david-allison-1 — APPROVED** [review]
LGTM, will add in a follow-up issue to add back the visual feedback

**Review by david-allison-1 — CHANGES_REQUESTED** [review]
Tests failing

**Comment by kuroni** [comment]
Ooops, totally did not expect the method to come up there :sweat:

**Review by david-allison-1 — APPROVED** [review]
Ditto previous: will need a follow-up issue to fix the visual regression

**Comment by kuroni** [comment]
> LGTM, will add in a follow-up issue to add back the visual feedback

Sorry, which visual feedback is this? I'm not totally clear.

**Comment by david-allison-1** [comment]
Hmm.... interesting. No regression here at all. I don't use gestures, but from the method name, I would have presumed that they would have animated the answer buttons

**Comment by david-allison-1** [comment]
@kuroni, apologies for the wasted effort, given [your comment](https://github.com/ankidroid/Anki-Android/issues/8583#issuecomment-817402056), I feel that a revert (#8584) is going to be the correct solution for now

**Comment by david-allison-1** [comment]
Will leave to the second reviewer to decide on the course of action

**Comment by kuroni** [comment]
@david-allison-1 I've implemented the suggestions from https://github.com/ankidroid/Anki-Android/issues/8583#issuecomment-817801198, the change brought back the feedback on my device. However I got unit tests failed here.

**Review by david-allison-1 — CHANGES_REQUESTED** [review]
The button still doesn't "feel" right to me: I had an instance (non-reproducible) where the highlighting wasn't removed.

The other issue is that pressing, dragging your finger off, and letting go of the button no longer stops the execution (and there doesn't seem to be an `ACTION_CANCEL` fired all the time).

As a proposed solution:

* Use a class which encapsulates both the touch and click listener
* on Touch (down), set the variable from the current card and return false to not consume the action
* on click, check the variable and perform the action
* Apply both the click and the touch listener to the view

I think that would be the best of both worlds.

Don't feel the need to do this. Totally optional, and I'd be happy to take it on (let me know!). Don't want to be changing scope of a PR, and this feels like a rewrite at this stage

**Comment by kuroni** [comment]
I think I'm up to the task; just give me a day or two.

**Comment by kuroni** [comment]
So after implementing for a while, I came into this problem where you cannot use the gesture for answer buttons. The reason is simple: we only set `mPrevCard` at `onTouch`, and when you use gesture you don't touch the ease layout.

**Comment by kuroni** [comment]
Please don't approve the current commit, there's a bug with the current logic.

Namely, suppose that you touch the button, but then moved your finger away from the button, that will trigger an `onTouch` event (which will set `mHasBeenTouched` to `true`), but will not trigger an `onClick` event. Now, if we let the card auto skip to the next card, and perform swipe, it will not register, because swiping only triggers the `onClick` event, which has both conditions on `mPrevCard` and `mHasBeenTouched` false.

I'm still trying to find a feasible action that can behave like `onTouch` with `ACTION_UP`, but is fired after `onClick` rather than before, but for now I cannot find anything like it ...

**Comment by kuroni** [comment]
Sorry, I just came back to this. I think I've covered the logic issue. @david-allison-1 can you take a look at the changes please?

File: AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.java [hunk 1]
```diff
@@ -429,53 +429,94 @@ public void run() {
...
-            // Perform intended action only if the button has been pressed for current card
-            if (event.getAction() == MotionEvent.ACTION_UP && mPrevCard == mCurrentCard) {
-                // Ignore what is most likely an accidental double-tap.
-                if (SystemClock.elapsedRealtime() - mLastClickTime < DOUBLE_TAP_IGNORE_THRESHOLD) {
-                    return false;
+                mHasBeenTouched = true;
+                mTouchX = event.getRawX();
+                mTouchY = event.getRawY();
+            } else if (event.getAction() == MotionEvent.ACTION_UP) {
+                float diffX = Math.abs(event.getRawX() - mTouchX);
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.java:462]: We could do with more comments here, particularly: explain the order of operations (touch before onClick) so a future developer understands why this logic exists

**Review by david-allison-1 — COMMENTED** [review]
Tested (& with keyboard) and all looks good.

Minor documentation nit and good to go

File: AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.java (hunk 1 above)
Comment (kuroni) [reply AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.java:466]: Resolved

**Review by david-allison-1 — APPROVED** [review]
LGTM! Thanks

File: AnkiDroid/src/test/java/com/ichi2/anki/AbstractFlashcardViewerTest.java [hunk 2]
```diff
@@ -384,6 +403,23 @@ public void testClozeWithRepeatedWords() {
         assertEquals("test, test, test2", nafv.contentForCloze(cloze2, 1));
     }
 
+    @Test
+    public void testCommandPerformsAnswerCard() {
+        // Regression for #8527/#8572
+        // Note: Couldn't get a spy working, so overriding the method
```
Comment (mikehardy) [inline AnkiDroid/src/test/java/com/ichi2/anki/AbstractFlashcardViewerTest.java:409]: I have these same difficulties sometimes, but getting a test in (even if it requires some hackery) is fantastic

**Comment by david-allison-1** [comment]
@mikehardy Are we good to go with this?

**Review by mikehardy — CHANGES_REQUESTED** [review]
This is one of our biggest items for release-2.15, thank you for tackling this!
My comments are important to me (of course! :-) ) but really trivial in general, super easy to fix I think. Looking forward to merging this :rocket:

File: AnkiDroid/src/test/java/com/ichi2/anki/AbstractFlashcardViewerTest.java [hunk 3]
```diff
@@ -384,6 +403,23 @@ public void testClozeWithRepeatedWords() {
...
+        NonAbstractFlashcardViewer viewer = getViewer();
+
+        assertThat("Displaying question", viewer.isDisplayingAnswer(), is(false));
+        viewer.executeCommand(ViewerCommand.COMMAND_FLIP_OR_ANSWER_BETTER_THAN_RECOMMENDED);
+
+        assertThat("Displaying answer", viewer.isDisplayingAnswer(), is(true));
+
+        viewer.executeCommand(ViewerCommand.COMMAND_FLIP_OR_ANSWER_BETTER_THAN_RECOMMENDED);
+
+        ShadowLooper.runUiThreadTasksIncludingDelayedTasks();
```
[the remark this answers is not in the document]
Comment (david-allison-1) [reply AnkiDroid/src/test/java/com/ichi2/anki/AbstractFlashcardViewerTest.java:419]: I added the test - we were performing a click action on a handler with a 20ms delay. This line should no longer be necessary.

[the remark this answers is not in the document]
Comment (mikehardy) [reply AnkiDroid/src/test/java/com/ichi2/anki/AbstractFlashcardViewerTest.java:419]: it still may be that `IncludingDelayedTasks` is superior than our current style, which is to say, harvesting the idea might be a positive even if a bit unrelated to this now

File: AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.java [hunk 4]
```diff
@@ -429,53 +429,98 @@ public void run() {
...
             displayCardAnswer();
         }
     };
 
-    private final View.OnTouchListener mSelectEaseHandler = new View.OnTouchListener() {
-        Card mPrevCard;
+
+    // Event handler for eases (answer buttons)
+    class SelectEaseHandler implements View.OnClickListener, View.OnTouchListener {
+        private static final int CLICK_ACTION_THRESHOLD = 200;
```
[the remark this answers is not in the document]
Comment (kuroni) [reply AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.java:445]: That is indeed the case. Thank you very much!

File: AnkiDroid/src/test/java/com/ichi2/anki/AbstractFlashcardViewerTest.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (kuroni) [reply AnkiDroid/src/test/java/com/ichi2/anki/AbstractFlashcardViewerTest.java:419]: I removed the line 419 from the file and the test passed. Should I go ahead and commit the removal, or do you have other suggestions (I'm not catching up with @mikehardy's latest comment so an explanation would be great)

[the remark this answers is not in the document]
Comment (mikehardy) [reply AnkiDroid/src/test/java/com/ichi2/anki/AbstractFlashcardViewerTest.java:419]: if you remove the line and it works, I would commit the removal of the line yes.
Sorry I was unclear, what I was trying (badly) to communicate was that this specific wait method looks different than our current style:

https://github.com/ankidroid/Anki-Android/blob/143d7bbf7c9aa87f842d44c240c07b7cb2bd9285/AnkiDroid/src/test/java/com/ichi2/anki/RobolectricTest.java#L229-L245

...and as a not-really-related item maybe we should add `ShadowLooper.runUiThreadTasksIncludingDelayedTasks()` to the RobolectricTest advance method(s).

I will peel this comment off to a separate issue for investigation and consider this resolved once you commit the removal of the line in this PR

Thanks @kuroni

**Review by mikehardy — APPROVED** [review]
Really close now - maintainer note is that this is completely ready for merge (I think) as soon as the one shadow looper wait call is removed from a test. Approving pending that

File: AnkiDroid/src/test/java/com/ichi2/anki/AbstractFlashcardViewerTest.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (kuroni) [reply AnkiDroid/src/test/java/com/ichi2/anki/AbstractFlashcardViewerTest.java]: Committed, thank you!

**Comment by mikehardy** [comment]
Hi there!

Just a friendly notice that we try to process OpenCollective payments monthly - it's time for April 2021 submissions

If you are interested in compensation for this work, the process with details is here:

https://github.com/ankidroid/Anki-Android/wiki/OpenCollective-Payment-Process#how-to-get-paid

(I only post one comment per person to avoid spamming you, regardless of the number of PRs merged, but this note applies to all PRs merged in April)

Thanks!

**Pull request merged**